File: week2/request1.py
def function1(name):
    people: dict[str, tuple[int, int]]={
        "悟空": (0, 0),
        "辛巴": (-3, 3),
        "貝吉塔": (-4, -1),
        "特南克斯": (1, -2),
        "佛利沙": (4, -1),
        "丁滿": (-1, 4)
    }
    section1 = ["悟空", "辛巴", "貝吉塔", "特南克斯"]
    section2 = ["佛利沙", "丁滿"]

    distance: dict[str, int] = {}
    
    def find_section(name):
        if name in section1:
            return 1
        elif name in section2:
            return 2
    target_section = find_section(name)
    extra_point = 2

    for n in people:
        if n == name:
            continue
        else:
            diff = abs(people[n][0] - people[name][0]) + abs(people[n][1] - people[name][1])
        if find_section(n) != target_section:
            diff = diff + extra_point
        distance[n] = diff
        
    value_max = max(distance.values())
    value_min = min(distance.values())
    max_key = []
    min_key = []
    for key, value in distance.items():
        if value == value_max:
            max_key.append(key)
        if value == value_min:
            min_key.append(key)
    print("最遠", "、".join(max_key), ";最近", "、".join(min_key))


print("================Task1================")
function1("辛巴")
function1("悟空")
function1("佛利沙")
function1("特南克斯")

File: week2/request1.js
function function1(name){
    people = [
        {"name": "悟空", "pos": [0, 0]},
        {"name": "辛巴", "pos": [-3, 3]},
        {"name": "貝吉塔", "pos": [-4, -1]},
        {"name": "特南克斯", "pos": [1, -2]},
        {"name": "佛利沙", "pos": [4, -1]},
        {"name": "丁滿", "pos": [-1, 4]},
    ];
    
    let section1 = ["悟空", "辛巴", "貝吉塔", "特南克斯"];
    let section2 = ["佛利沙", "丁滿"];
    // calculate the differences of input name and each name
    for (let i = 0; i < people.length; i++){
        if (people[i].name == name){
            POS = people[i].pos;
            break;
        }
    }
    // check section
    function section(name){
        for (let i = 0; i < section1.length; i++){
            if (section1[i] === name){
                return 1;
            }
        }
        for (let i = 0; i < section2.length; i++){
            return 2;
        }
        return console.log("Not Found");
    }

    let target_section = section(name);
    let new_list = [];
    for (let i = 0; i < people.length; i++){
        if (people[i].name == name){
            continue;
        }
        else if (section(people[i].name) === target_section){
            distance_raw = Math.abs(people[i].pos[0] - POS[0]) + Math.abs(people[i].pos[1] - POS[1]);
        }
        else if (section(people[i].name) != target_section){
            distance_raw = Math.abs(people[i].pos[0] - POS[0]) + Math.abs(people[i].pos[1] - POS[1]) + 2;
        }
        new_list.push({[people[i].name] : distance_raw})
    }

    let min = Infinity;
    let min_list = [];
    for (o of new_list){
        [name, value] = Object.entries(o)[0]
        if (value < min){
            min = value;
            min_list = [name];
        }
        else if (value === min){
            min_list.push(name);
        }
    }
    let max = -Infinity;
    let max_list = [];
    for (o of new_list){
        [name, value] = Object.entries(o)[0];
        if (value > max){
            max = value;
            max_list = [name];
        }
        else if(value === max){
            max_list.push(name);
        }
    }
    console.log("最遠", max_list.join("、"), "; 最近", min_list.join("、"));

    // check every name if they are  in the same section with input name

    // find the farest and closest one and print
    
}

console.log("================Task1================")
function1("辛巴")
function1("悟空")
function1("佛利沙")
function1("特南克斯")
